signal.go

```diff
@@ -0,0 +1,114 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package concert
+
+import (
+	"errors"
+	"sync"
+)
+
+// OnceSignaler provides a channel that can only be closed once.
+// In addition to the channel one can install callbacks to be executed
+// if the signal is triggered.
+// Once triggered all further close attempts will be ignored.
+//
+// The zero value is not valid. NewOnceSignaler must be used to create an
+// instance backed by a channel.
+type OnceSignaler struct {
+	once sync.Once
+	mu   sync.Mutex
+	ch   chan struct{}
+	fn   func()
+}
+
+var Canceled = errors.New("cancelled")
```

Comment:
Maybe renamed to `ErrCancelled`? It's probably missing a comment before, I don't think we have hound or govet running on this repository yet.

Reply:
I used Canceled, as this is more similar to `context.Canceled`... maybe it would be better to reuse `context.Context` and define this variable as `var Canceled = context.Canceled`...

Reply:
I think it's good.